google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbortedTest.java

```diff
@@ -201,63 +251,77 @@ public void testAbortedDuringRetryOfFailedQueryAsFirstStatement() {
-    final Statement invalidStatement = Statement.of("INSERT INTO FOO");
+    final Iterable<Statement> invalidStatements =
+        ImmutableList.of(
+            Statement.of("INSERT INTO FOO"), Statement.of("INSERT INTO FOO THEN RETURN *"));
     StatusRuntimeException notFound =
         Status.NOT_FOUND.withDescription("Table not found").asRuntimeException();
-    mockSpanner.putStatementResult(StatementResult.exception(invalidStatement, notFound));
-    try (ITConnection connection =
-        createConnection(createAbortFirstRetryListener(invalidStatement, notFound))) {
-      try {
-        connection.executeBatchUpdate(Collections.singletonList(invalidStatement));
-        fail("missing expected exception");
-      } catch (SpannerException e) {
-        assertThat(e.getErrorCode()).isEqualTo(ErrorCode.NOT_FOUND);
+    for (Statement invalidStatement : invalidStatements) {
+      mockSpanner.putStatementResult(StatementResult.exception(invalidStatement, notFound));
+      try (ITConnection connection =
+          createConnection(createAbortFirstRetryListener(invalidStatement, notFound))) {
+        try {
+          connection.executeBatchUpdate(Collections.singletonList(invalidStatement));
+          fail("missing expected exception");
+        } catch (SpannerException e) {
+          assertThat(e.getErrorCode()).isEqualTo(ErrorCode.NOT_FOUND);
+        }
+        connection.execute(INSERT_STATEMENT);
+        // Force an abort and retry.
+        mockSpanner.abortNextStatement();
+        connection.commit();
       }
-      connection.execute(INSERT_STATEMENT);
-      // Force an abort and retry.
-      mockSpanner.abortNextStatement();
-      connection.commit();
+      assertThat(mockSpanner.countRequestsOfType(CommitRequest.class)).isEqualTo(2);
+      assertThat(mockSpanner.countRequestsOfType(ExecuteBatchDmlRequest.class)).isEqualTo(6);
+      mockSpanner.clearRequests();
     }
-    assertThat(mockSpanner.countRequestsOfType(CommitRequest.class)).isEqualTo(2);
-    assertThat(mockSpanner.countRequestsOfType(ExecuteBatchDmlRequest.class)).isEqualTo(6);
   }
 
   @Test
   public void testRetryUsesTags() {
     mockSpanner.putStatementResult(
         StatementResult.query(SELECT_COUNT_STATEMENT, SELECT_COUNT_RESULTSET_BEFORE_INSERT));
     mockSpanner.putStatementResult(StatementResult.update(INSERT_STATEMENT, UPDATE_COUNT));
+    mockSpanner.putStatementResult(
+        StatementResult.updateReturning(INSERT_RETURNING_STATEMENT, UPDATE_RETURNING_RESULTSET));
     try (ITConnection connection = createConnection()) {
       connection.setTransactionTag("transaction-tag");
       connection.setStatementTag("statement-tag");
       connection.executeUpdate(INSERT_STATEMENT);
       connection.setStatementTag("statement-tag");
-      connection.executeBatchUpdate(Collections.singleton(INSERT_STATEMENT));
+      connection.executeBatchUpdate(ImmutableList.of(INSERT_STATEMENT, INSERT_RETURNING_STATEMENT));
```

Comment:
An interesting question here is whether the client should already throw an exception when you try to execute it, or whether we should rely on the backend doing so (at least, as far as I can remember, `BatchDml` does not support returning clauses).

Reply:
DML with Returning clause over `BatchDML` is allowed. Have added corresponding integration tests for `executeBatchDml` and `executeBatchDmlAsync` in ITDmlReturningTest in this PR.